fuse_core/include/fuse_core/time.h

```diff
@@ -0,0 +1,185 @@
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+
+
+/*
+ * This file provides a sane configuration for a clock to manage order of events.
+ * This file exists to help with a migration from fuse_core::TimeStamp to rclcpp::time
+ * Priority is given to C++ std::chrono
+ * emphasis will be placed on interoperability with fuse_core::TimeStamp and rclcpp::time
+ *
+ * This will likely be extended to distinguish between:
+ *   times of events being optimised
+ *   events generated by the running of the optimiser (log throttling, callbacks etc)
+ *
+ * The clock class will likely be overridden to prevent use of now(),
+ *   emphasising that the optimiser is not strictly realtime and the epoch is probably invalid
+ *
+ */
+
+#ifndef FUSE_CORE_TIME_H
+#define FUSE_CORE_TIME_H
+
+#include <chrono>
+#include <iostream>
+#include <stdexcept>
+
+#include "rclcpp/contexts/default_context.hpp"
+#include <rclcpp/clock.hpp>
+#include <rclcpp/duration.hpp>
+#include <rclcpp/time.hpp>
+
+
+namespace fuse_core
+{
+using Duration = std::chrono::nanoseconds;  // rclcpp uses nanosecond durations
+
+/**
+ * @brief Implements a Time class that supports interoperability between rclcpp and std::chrono time
+ *
+ * The class is implemented as a thin wrapper around rclcpp::Time. This allows us to take advantage
+ * of the operator overloads and clock type tracking of the parent class. Including the consistency
+ * checks for clock types when doing time arithmetic.
+ *
+ * To support this, the internal representation of time is an rclcpp::Time object and conversion
+ * functions and constructors are written to create this object out of primitive, rclcpp::Time and
+ * std::chrono time objects.
+ *
+ * A significant deviation from rclcpp:Time is the introduction of the default constructor that
+ * explicitly sets the clock type to RCL_CLOCK_UNINITIALIZED. This is done to differentiate between
+ * zero-initializations of time, and an uninitialized time.
```

Comment:
Do we need `fuse_core::TimeStamp`? I don't think the `RCL_CLOCK_UNINITIALIZED` in the constructor is needed anymore since 0 on any clock type will be invalid (for better or worse).\

Any conversions to chrono types can be done in free functions, but I'm not sure converting to `std::chrono` types is a good idea. There's no guarantee that the clock used to get the ROS time is the same clock used by `std::chrono`. For example, for steady time `rcutils` is using [`CLOCK_MONOTONIC_RAW`](https://github.com/ros2/rcutils/blob/f69678ae6b9126895482b57b0cf6c47862385e03/src/time_unix.c#L90) while [gcc's chrono implentation is using `CLOCK_MONOTONIC`](https://github.com/gcc-mirror/gcc/blob/32a06ce38a38bf37db468f0e6c83520fcc221534/libstdc%2B%2B-v3/src/c%2B%2B11/chrono.cc#L85). `std::chrono::time_point<>` returned by both might give incorrect comparisons because [one is subject to NTP adjustments](https://linux.die.net/man/2/clock_gettime). The same could happen with system time depending on which C++ standard library implementation is used.

Reply:
`fuse_core::TimeStamp` was mostly part of a plan to support backports to Fuse for ROS1
`std::chrono` isn't required, as long as there's enough feature parity that Fuse doesn't diverge substantially between ROS1 and ROS2 branches

I do worry about applications crashing when a (valid) time zero arrives

Reply:
I think we're currently at a point where we can indeed use `rclcpp::Time` :thinking: 
The only other reason I can think of for wrapping the time class would be to add metadata, which isn't a concern at the moment.

For posterity though, I've saved the current state in a gist: https://gist.github.com/methylDragon/ba1a39f619d7f9ca22dfc9fd49e51b40

As for crashing when time zero arrives, the only time that could occur is if a system clock is reporting time 0 on its epoch (meaning it'd be in 1970, 1 Jan), or if we have time 0 in a simulation. The former is unlikely to happen, and the latter can be guarded against with https://github.com/methylDragon/fuse/blob/rolling-time/fuse_optimizers/src/optimizer.cpp#L108 , so I tihnk it's fairly safe. In either case, those timestamps would be considered invalid.

I'll also update all Time instantiations of (0, 0) to (0, 1), to reflect the time validity. Change is pending.

Reply:
Again, the point of `fuse_core::TimeStamp` was for ROS1 backports, and I still think that's an important usage for a wrapper class, even if it's just a typedef.
Time is currently the only ROS dependency in user-written `fuse_variables`. If you wrap time, migration of variable definitions between versions of ROS becomes trivial

I also want to emphatically dispute "unlikely" because ROS2 uses signed 32 bit integers on the system epoch, which overflows to negative values and stays there in a little less than 16 years from now.
This will never be a problem in ROS1 because nobody will accumulate 60 years of up-time on any ROS1 roscore process
The embedded systems world has conventions where we use overflow-tolerant arithmetic, but it strictly requires that the actual time values are never used as logical flags.

Reply:
> Again, the point of fuse_core::TimeStamp was for ROS1 backports, and I still think that's an important usage for a wrapper class, even if it's just a typedef.
Time is currently the only ROS dependency in user-written fuse_variables. If you wrap time, migration of variable definitions between versions of ROS becomes trivial

The typedef will need to be done to both ROS 1 and ROS 2 branches to be useful. Supporting ROS 1 is out of scope for our effort, but I don't see a reason against doing so as an enhancement to both branches later. In the meantime user-written `fuse_variables` can support both ROS versions with `#ifdef RCLCPP_VERSION_MAJOR` and `typedef`. That's not convenient, but I would guess downstream code probably already has some ROS-awareness at the build system level because of catkin.

> I also want to emphatically dispute "unlikely" because ROS2 uses signed 32 bit integers on the system epoch, which overflows to negative values and stays there in a little less than 16 years from now.

This is a larger issue than we can solve in the `fuse` repo. ROS 2 will probably follow whatever OMG DDS does https://issues.omg.org/issues/DDS15-13